Design-Docs/standardizing-cluster-infrastructure.md

```diff
@@ -0,0 +1,277 @@
+## Risks and Mitigations
+
+The risks are the same as every standardization has. Cloud providers might not be satisfied by the selection of cluster stacks that SCS chose. As soon as one wish is granted, others come and want to have their ideas implemented in SCS as well. In the end, there is no standardization and a useless certification of SCS infrastructure.
+
+## Design details
+
+As described above, we want to achieve two things: standardizing clusters and making applications independent of the cluster infrastructure itself. By separating these two layers, we achieve a clear definition of responsibilities and compatibility of different cloud providers.
+
+In this way, standardized but also configurable Kubernetes clusters can be created in self-service - also via GitOps.
+
+### Proposal 1: Using cluster stacks
+
+There are various components playing a role in defining cluster infrastructure. We consider the following three technical layers:
+
+1. Pre-cluster state - node image
+2. Cluster state - Cluster API and provider implementation
+3. Kubernetes core applications
+
+All of these layers depend on each other, such that it is not possible to freely select and exchange components. This is why we introduce the cluster stack which consists of a selection of sub-systems and configurations that follow a certain goal.
+
+For example, there can be a stack for Kubernetes version 1.22 and one for 1.23. There can also be stacks that use a node image based on Ubuntu and others with Fedora. These cluster stacks have to be implemented by the providers, as there are certain details, e.g. related to the node image, that have to be adapted to the respective provider.
+
+The process of building and implementing these cluster stacks has to be well designed, as there are new patch versions every couple of weeks. Providers should be able to implement cluster stacks within a few days. E2E tests play a big role in automating work and ensuring that everything works.
+
+#### What exactly is a cluster stack?
+
+First, SCS defines a stack and decides that it should use a certain Kubernetes version, that it has a certain node image, etc. Based on these decisions, the stack is defined in yaml files using some templating language (e.g. Helm/Helmfile or Jsonnet). These yaml files can be seen as a framework that is implemented by the providers. Some parts of the cluster stack configuration have to be adapted by the cloud providers. Good documentation from SCS ensures that implementing a cluster stack is straightforward. A templating language is used, so that cluster operators can specify certain configurations of the cluster stacks that have not been fixed. For example, the configuration of OpenID Connect.
+
+#### What components does a cluster stack have? What has to be considered when defining a stack?
+
+Based on the aforementioned layers, we want to make a list of all technical details, cluster components, and things to consider.
+
+##### Layer 1: Pre-cluster state
```

Comment:
Some of these things in the node look like highly specific implementation detail to me.

I think we should define rules on how to decide if something should be part of the node.
For example, if you are able to add a piece of software via helm or Kubernetes YAML files only there should be no need to implement this at the node image level.

There are also details like `cgroupfs` vs `systemd` which IIRC should not have any impact to the cluster operator.
At gridscale, we switched to `systemd` because we have better stability in high memory pressure situations with this config. However, this could potentially be different on other systems.

In general, we should define a goal here but not define the specific piece of software.

Reply:
The choice of cgroup handling is certainly visible, at least in some use-cases: https://gitlab.com/yaook/operator/-/issues/345

Might be that this specific one should be out of scope…